apm-agent-common/src/main/java/co/elastic/apm/agent/common/util/ResourceExtractionUtil.java

```diff
@@ -18,98 +18,127 @@
+            // to multiple JVMs, each running under a different user. Hashing makes the name path-friendly.
+            String userHash = hash(currentUserPrincipal.getName());
+            // to guard against re-using previous versions
+            String resourceHash = hash(ResourceExtractionUtil.class.getResourceAsStream("/" + resource));
 
-            File tempFile = new File(directory, prefix + "-" + userHash + resourceHash + suffix);
-            if (!tempFile.exists()) {
-                try (FileOutputStream out = new FileOutputStream(tempFile)) {
-                    FileChannel channel = out.getChannel();
-                    // If multiple JVM start on same compute, they can write in same file
-                    // and this file will be corrupted.
-                    try (FileLock ignored = channel.lock()) {
-                        if (tempFile.length() == 0) {
-                            byte[] buffer = new byte[1024];
-                            for (int length; (length = resourceStream.read(buffer)) != -1; ) {
-                                out.write(buffer, 0, length);
-                            }
+            Path tempFile = directory.resolve(prefix + "-" + userHash.substring(0, 32) + "-" + resourceHash.substring(0, 32) + suffix);
+            try {
+                FileAttribute<?>[] attr;
+                if (tempFile.getFileSystem().supportedFileAttributeViews().contains("posix")) {
+                    attr = new FileAttribute[]{PosixFilePermissions.asFileAttribute(EnumSet.of(OWNER_WRITE, OWNER_READ))};
+                } else {
+                    attr = new FileAttribute[0];
+                }
+                try (FileChannel channel = FileChannel.open(tempFile, EnumSet.of(CREATE_NEW, WRITE), attr)) {
+                    // make other JVM instances wait until fully written
+                    try (FileLock writeLock = channel.lock()) {
+                        channel.transferFrom(Channels.newChannel(resourceStream), 0, Long.MAX_VALUE);
+                    }
+                }
+            } catch (FileAlreadyExistsException e) {
+                try (FileChannel channel = FileChannel.open(tempFile, READ, NOFOLLOW_LINKS)) {
+                    // wait until other JVM instances have fully written the file
+                    // multiple JVMs can read the file at the same time
+                    try (FileLock readLock = channel.lock(0, Long.MAX_VALUE, true)) {
+                        if (!hash(Files.newInputStream(tempFile)).equals(resourceHash)) {
+                            throw new IllegalStateException("Invalid checksum of " + tempFile + ". Please delete this file.");
+                        } else if (!Files.getOwner(tempFile).equals(currentUserPrincipal)) {
+                            throw new IllegalStateException("File " + tempFile + " is not owned by '" + currentUserPrincipal.getName() + "'. Please delete this file.");
                         }
                     }
                 }
```

Comment:
[proposal] This is a `synchronized` method, I think a clearer path is to start with checking `java.nio.file.Files#exists()` and if so do all that for the internal JVM synchronization. 
You can extract this to a different method and do the same check and handling after getting the `writeLock` for the cross-JVMs synchronization.
I think it's clearer than relying on getting a `FileAlreadyExistsException`.

Reply:
It's actually important to rely on the exception as `Files#exists` is vulnerable to check-then-act race conditions.

`FileChannel.open`, can atomically create a new file with a specific set of file attributes. It's guaranteed that if it returns without throwing an exception, no other user has created a file with the same name.

Reply:
OK